### PR TITLE
Support multiple opensearch versions 2

### PR DESCRIPTION
**Pull request by jeffmclean-cs**
An alternative to https://github.com/CrowdStrike/opensearchtools/pull/11 which I think is an improvement because it:
- lets a the consumer chose to either use the OpenSearch 2 specific stuff without the extra abstraction of using a version-agnostic mechanism 
- still allows consumers who wish to use a version-agnostic mechanism to do so which should make going from OpenSearch 2 to a future OpenSearch 3, etc relatively seamless and also do things like allow dual-writes (ex to OpenSearch 2 and OpenSearch 3).

Loosely put, it also keeps the Core Maintainers goal about "let's make a nice abstraction to facilitate different versions, etc" a bit more cleanly separated Marshall's more immediately useful OS2-specific code and allows to consumer to chose to use the simpler OS2 specific API or the more elaborate version agnostic API. This is more similar to the orig than https://github.com/CrowdStrike/opensearchtools/pull/11 is.

### CODE REVIEW

File: mget.go [hunk 1]
```diff
@@ -25,20 +27,19 @@ type MGetRequest struct {
 	Docs []RoutableDoc
 }
 
-// NewMGetRequest instantiates an empty MGetRequest.
-// An empty MGetRequest is executable but will return zero documents because zero have been requested.
```
Comment (dylan-bourque) [inline mget.go:29]: Keep this extra context in the godoc comment

File: osv2/executor.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import (
+	"context"
+
+	"github.com/opensearch-project/opensearch-go/v2"
+
+	"github.com/CrowdStrike/opensearchtools"
+)
+
+// OSv2Executor is an executor for OpenSearch 2.
+type OSv2Executor struct {
```
Comment (dylan-bourque) [inline osv2/executor.go:12]: Call this `Executor` since it will be referenced as `osv2.Executor` on the consumer side.  The `OSv2` prefix stutters.

**Review by dylan-bourque — COMMENTED** [review]
Had a couple of suggestions but I think this has potential.  It does feel a bit cleaner than #11 while accomplishing the same goals.

File: mget.go (hunk 1 above)
Comment (jeffmclean-cs) [reply mget.go:29]: done

File: osv2/executor.go (hunk 2 above)
Comment (jeffmclean-cs) [reply osv2/executor.go]: Done

File: mget.go [hunk 3]
```diff
@@ -1,15 +1,17 @@
...
-	"bytes"
 	"context"
 	"encoding/json"
 	"net/http"
-
-	"github.com/opensearch-project/opensearch-go/v2"
-	"github.com/opensearch-project/opensearch-go/v2/opensearchapi"
 )
 
+// MGet defines a method which knows how to make an OpenSearch multiple document get.
```
Comment (cs-marshall) [inline mget.go:9]: I would change this comment, as it doesn't quite line up with OpenSearch verbiage.
```
// MGet defines a method which knows how to make an OpenSearch [Multi-get] request.
// It should be implemented by a version-specific executor.
//
// [Multi-get]: https://opensearch.org/docs/latest/api-reference/document-apis/multi-get/
```

File: mget.go [hunk 4]
```diff
@@ -1,15 +1,17 @@
...
+// It should be implemented by a version-specific executor.
+type MGet interface {
+	MGet(ctx context.Context, req *MGetRequest) (*MGetResponse, error)
+}
+
 // MGetRequest wraps the functionality of [opensearchapi.MgetRequest] by supporting request body creation.
 // We can perform an MGetRequest as simply as:
```
Comment (cs-marshall) [inline mget.go:16]: If this is the pattern we go with, this doc comment sure be updated to explicitly call out that it's a union of all OpenSearch version requests and that individual executors will listen to and ignore desired fields.

Comment (cs-marshall) [reply mget.go:16]: Same for the `MGetResponse`. Maybe link and list the supported version implementations.

File: osv2/mget.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+	Version     int             `json:"_version,omitempty"`
+	SeqNo       int             `json:"_seq_no,omitempty"`
+	PrimaryTerm int             `json:"_primary_term,omitempty"`
+	Found       bool            `json:"found,omitempty"`
+	Source      json.RawMessage `json:"_source,omitempty"`
+	Error       error           `json:"-"`
+}
+
+// ToModel converts this instance of an [MGetResult] into an [opensearchtools.MGetResult].
+func (r *MGetResult) ToModel() opensearchtools.MGetResult {
```
Comment (cs-marshall) [inline osv2/mget.go:129]: There's something that feels lacking to me with just `ToModel`, but I'm not sure if I have an actual answer. Maybe too generic? `ToDomainModel` feels slightly better to me, but maybe domain and model are redundant.

File: osv2/mget_test.go [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+		Version     int
+		SeqNo       int
+		PrimaryTerm int
+		Found       bool
+		Source      json.RawMessage
+		Error       error
+	}
+	tests := []struct {
+		name   string
+		fields fields
```
Comment (cs-marshall) [inline osv2/mget_test.go:159]: Do you like the auto generated `fields` pattern? I personally don't and would just have a `target MGetResult` attribute instead.

File: osv2/mget_test.go [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+							SeqNo:       220,
+							PrimaryTerm: 30,
+							Found:       false,
+							Source:      []byte{},
+							Error:       nil,
+						},
+					},
+				},
+			},
+			want: &opensearchtools.MGetResponse{
```
Comment (cs-marshall) [inline osv2/mget_test.go:266]: let's add a test for an error case where no results are returned.

**Review by cs-marshall — CHANGES_REQUESTED** [review]

File: osv2/mget_test.go (hunk 6 above)
Comment (jeffmclean-cs) [reply osv2/mget_test.go:159]: Ya, I agree. It seems pointless.

File: osv2/mget_test.go (hunk 7 above)
Comment (jeffmclean-cs) [reply osv2/mget_test.go:266]: done

File: osv2/mget_test.go (hunk 6 above)
Comment (jeffmclean-cs) [reply osv2/mget_test.go:159]: addressed this in another commit

File: mget.go (hunk 3 above)
Comment (jeffmclean-cs) [reply mget.go:9]: Thanks - done

File: osv2/mget.go (hunk 5 above)
Comment (jeffmclean-cs) [reply osv2/mget.go:129]: Definitely up for a better name, though I agree ToDomainModel doesn't really tell you any more than ToModel.

File: mget.go (hunk 4 above)
Comment (jeffmclean-cs) [reply mget.go]: @cs-marshall I updated the doc comments for this (MGetRequest) as well as MGetResponse and MGetResult. LMK what you think.

File: mget.go [hunk 8]
```diff
@@ -51,87 +65,33 @@ func (m *MGetRequest) SetIndex(index string) *MGetRequest {
...
-		"docs": docs,
-	}
-
-	return json.Marshal(source)
-}
-
-// Do executes the Multi-Get MGetRequest using the provided opensearch.Client.
-// If the request is executed successfully, then a MGetResponse with MGetResults will be returned.
-// An error can be returned if
+// MGetResponse is a domain model response type for Multi-Get for all supported OpenSearch versions.
```
Comment (cs-marshall) [inline mget.go:68]: domain model **union** response type

File: mget.go [hunk 9]
```diff
@@ -51,87 +65,33 @@ func (m *MGetRequest) SetIndex(index string) *MGetRequest {
...
-	StatusCode int          `json:"-"`
-	Header     http.Header  `json:"-"`
-	Docs       []MGetResult `json:"docs,omitempty"`
+	Header     http.Header
+	StatusCode int
+	Docs       []MGetResult
 }
 
-// MGetResult is the individual result for each requested item.
+// MGetResult is a domain model type representing an individual document result in for a request item
```
Comment (cs-marshall) [inline mget.go:79]: **union** type

File: osv2/mget.go [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+	"encoding/json"
+	"net/http"
+
+	"github.com/opensearch-project/opensearch-go/v2"
+	"github.com/opensearch-project/opensearch-go/v2/opensearchapi"
+
+	"github.com/CrowdStrike/opensearchtools"
+)
+
+// MGetRequest is a marshalable form of [opensearchtools.MGetRequest] specific to the opensearchapi.MgetRequest in OpenSearch v2.
```
Comment (cs-marshall) [inline osv2/mget.go:15]: do we want to link the 2.x documentation for multi-get?

File: osv2/mget.go (hunk 5 above)
Comment (cs-marshall) [reply osv2/mget.go:135]: We can punt it until we think of a better name.

File: osv2/mget.go [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+type MGetRequest struct {
```
Comment (cs-marshall) [inline osv2/mget.go:16]: Should we also have `Add` and `AddDocs` for this? So if a consumer just wants to live if OSV2 they can? Or is that a bit of overkill?

**Review by cs-marshall — APPROVED** [review]

File: mget.go (hunk 8 above)
Comment (jeffmclean-cs) [reply mget.go:68]: updated

File: mget.go (hunk 9 above)
Comment (jeffmclean-cs) [reply mget.go:79]: updated

File: osv2/mget.go (hunk 10 above)
Comment (jeffmclean-cs) [reply osv2/mget.go:15]: I can't seem to get a link to 2.x - I can either do a specific version or `latest`. I could put in 2.5, but then that would be out of date when they add 2.6. I could link to `latest`, but then that will be wrong when they add version 3.

File: osv2/mget.go (hunk 11 above)
Comment (jeffmclean-cs) [reply osv2/mget.go:16]: Yes, definitely. I'll add them. Thanks.

Comment (jeffmclean-cs) [reply osv2/mget.go:18]: @cs-marshall I added them and also added a the constructor for MGetRequest to osv2/MGetRequest.
And also renamed `SetIndex` to `WithIndex` (same idea as in your PR).

File: osv2/mget.go (hunk 10 above)
Comment (dylan-bourque) [reply osv2/mget.go:15]: I agree with Jeff on this one.  There's no URL to the docs for v2.x so we shouldn't try to DIY it.

**Review by dylan-bourque — APPROVED** [review]
I think this is a solid starting point.  We can iterate from here as we build things out.

File: osv2/mget.go (hunk 10 above)
Comment (cs-marshall) [reply osv2/mget.go:15]: fair enough

**Pull request merged**